File: clients/ios/Lextures/Core/LMS/PlannerLogic.swift
import Foundation

// MARK: - Models

enum StudentTodoKind: String, Hashable {
    case dueItem
    case notebookTask
}

enum StudentTodoCompletion: String, Hashable {
    case open
    case submitted
    case completed
}

enum StudentTodoBucket: String, CaseIterable, Identifiable {
    case overdue
    case today
    case thisWeek
    case later

    var id: String { rawValue }
}

enum PlannerCalendarEventKind: String, Hashable {
    case assignment
    case quiz
    case contentPage
    case notebookTask
    case academic
    case officeHours
}

struct StudentTodoItem: Identifiable, Hashable {
    var id: String { key }
    let key: String
    let kind: StudentTodoKind
    let title: String
    let courseCode: String
    let courseTitle: String
    let dueAt: Date?
    let structureKind: String?
    let structureItemId: String?
    let notebookPageId: String?
    let notebookTaskId: String?
    var completion: StudentTodoCompletion

    var isCompleted: Bool { completion == .completed || completion == .submitted }
}

struct PlannerCalendarEvent: Identifiable, Hashable {
    var id: String
    let title: String
    let courseCode: String?
    let courseTitle: String?
    let startsAt: Date
    let endsAt: Date?
    let allDay: Bool
    let kind: PlannerCalendarEventKind
    let structureKind: String?
    let structureItemId: String?
    let notebookPageId: String?
    let officeHoursSlotId: String?
    let meetingId: String?

    init(
        id: String,
        title: String,
        courseCode: String?,
        courseTitle: String?,
        startsAt: Date,
        endsAt: Date?,
        allDay: Bool,
        kind: PlannerCalendarEventKind,
        structureKind: String?,
        structureItemId: String?,
        notebookPageId: String?,
        officeHoursSlotId: String? = nil,
        meetingId: String? = nil
    ) {
        self.id = id
        self.title = title
        self.courseCode = courseCode
        self.courseTitle = courseTitle
        self.startsAt = startsAt
        self.endsAt = endsAt
        self.allDay = allDay
        self.kind = kind
        self.structureKind = structureKind
        self.structureItemId = structureItemId
        self.notebookPageId = notebookPageId
        self.officeHoursSlotId = officeHoursSlotId
        self.meetingId = meetingId
    }
}

struct PlannerCourseFilter: Hashable, Identifiable {
    var id: String { courseCode }
    let courseCode: String
    let title: String
}

/// Cached planner payload for offline reads.
struct PlannerSnapshot: Codable {
    var fetchedAt: Date
    var todos: [CachedStudentTodoItem]
    var events: [CachedPlannerCalendarEvent]
}

struct CachedStudentTodoItem: Codable, Hashable {
    var key: String
    var kind: String
    var title: String
    var courseCode: String
    var courseTitle: String
    var dueAt: String?
    var structureKind: String?
    var structureItemId: String?
    var notebookPageId: String?
    var notebookTaskId: String?
    var completion: String
}

struct CachedPlannerCalendarEvent: Codable, Hashable {
    var id: String
    var title: String
    var courseCode: String?
    var courseTitle: String?
    var startsAt: String
    var endsAt: String?
    var allDay: Bool
    var kind: String
    var structureKind: String?
    var structureItemId: String?
    var notebookPageId: String?
    var officeHoursSlotId: String?
    var meetingId: String?
}

enum DueReminderLeadTime: Int, CaseIterable, Identifiable {
    case none = 0
    case fifteenMinutes = 15
    case oneHour = 60
    case oneDay = 1440

    var id: Int { rawValue }

    var labelKey: String {
        switch self {
        case .none: return "mobile.planner.reminder.none"
        case .fifteenMinutes: return "mobile.planner.reminder.15m"
        case .oneHour: return "mobile.planner.reminder.1h"
        case .oneDay: return "mobile.planner.reminder.1d"
        }
    }
}

// MARK: - Collection

enum PlannerLogic {
    static func dueItemKey(courseCode: String, itemId: String) -> String {
        "due:\(courseCode):\(itemId)"
    }

    static func notebookTaskKey(taskId: String) -> String {
        "notebook:\(taskId)"
    }

    static func collectTodos(
        studentCourses: [CourseSummary],
        structureByCourseCode: [String: [CourseStructureItem]],
        notebookTasks: [NotebookTask],
        gradesByCourseCode: [String: MyGradesResponse]
    ) -> [StudentTodoItem] {
        let courseTitles = Dictionary(uniqueKeysWithValues: studentCourses.map { ($0.courseCode, $0.displayTitle) })
        let studentCodes = Set(studentCourses.map(\.courseCode))
        var items: [StudentTodoItem] = []

        for task in notebookTasks where !task.completed {
            if task.courseCode != "__global__", !studentCodes.contains(task.courseCode) { continue }
            let title = task.taskText.trimmingCharacters(in: .whitespacesAndNewlines)
            items.append(
                StudentTodoItem(
                    key: notebookTaskKey(taskId: task.id),
                    kind: .notebookTask,
                    title: title.isEmpty ? "Untitled task" : title,
                    courseCode: task.courseCode,
                    courseTitle: courseTitle(for: task.courseCode, titles: courseTitles),
                    dueAt: LMSDates.parse(task.dueAt),
                    structureKind: nil,
                    structureItemId: nil,
                    notebookPageId: task.notebookPageId,
                    notebookTaskId: task.id,
                    completion: task.completed ? .completed : .open
                )
            )
        }

        for course in studentCourses where course.isCalendarEnabled {
            let structure = structureByCourseCode[course.courseCode] ?? []
            let grades = gradesByCourseCode[course.courseCode]
            for row in structure where isDueStructureItem(row) {
                let completion = completionStatus(
                    itemId: row.id,
                    kind: row.kind,
                    grades: grades
                )
                items.append(
                    StudentTodoItem(
                        key: dueItemKey(courseCode: course.courseCode, itemId: row.id),
                        kind: .dueItem,
                        title: row.title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
                            ? "Untitled"
                            : row.title.trimmingCharacters(in: .whitespacesAndNewlines),
                        courseCode: course.courseCode,
                        courseTitle: course.displayTitle,
                        dueAt: LMSDates.parse(row.dueAt),
                        structureKind: row.kind,
                        structureItemId: row.id,
                        notebookPageId: nil,
                        notebookTaskId: nil,
                        completion: completion
                    )
                )
            }
        }

        return items.sorted { lhs, rhs in
            switch (lhs.dueAt, rhs.dueAt) {
            case let (left?, right?): return left < right
            case (nil, _?): return false
            case (_?, nil): return true
            case (nil, nil): return lhs.title.localizedCaseInsensitiveCompare(rhs.title) == .orderedAscending
            }
        }
    }

    static func bucketTodos(_ items: [StudentTodoItem], now: Date = Date()) -> [StudentTodoBucket: [StudentTodoItem]] {
        var calendar = Calendar.current
        calendar.firstWeekday = 2
        let startOfToday = calendar.startOfDay(for: now)
        let endOfToday = calendar.date(byAdding: DateComponents(day: 1, second: -1), to: startOfToday) ?? now
        let weekStart = calendar.dateInterval(of: .weekOfYear, for: now)?.start ?? startOfToday
        let weekEnd = calendar.date(byAdding: DateComponents(day: 7, second: -1), to: weekStart) ?? now

        var buckets = Dictionary(uniqueKeysWithValues: StudentTodoBucket.allCases.map { ($0, [StudentTodoItem]()) })
        for item in items {
            guard let due = item.dueAt else {
                buckets[.later, default: []].append(item)
                continue
            }
            if due < startOfToday {
                buckets[.overdue, default: []].append(item)
            } else if due <= endOfToday {
                buckets[.today, default: []].append(item)
            } else if due <= weekEnd {
                buckets[.thisWeek, default: []].append(item)
            } else {
                buckets[.later, default: []].append(item)
            }
        }
        return buckets
    }

    static func collectCalendarEvents(
        studentCourses: [CourseSummary],
        structureByCourseCode: [String: [CourseStructureItem]],
        notebookTasks: [NotebookTask],
        academicEvents: [AcademicCalendarEvent],
        officeHoursByCourseCode: [String: OfficeHoursAvailability] = [:]
    ) -> [PlannerCalendarEvent] {
        var events: [PlannerCalendarEvent] = []
        let courseTitles = Dictionary(uniqueKeysWithValues: studentCourses.map { ($0.courseCode, $0.displayTitle) })

        for course in studentCourses where course.isCalendarEnabled {
            let structure = structureByCourseCode[course.courseCode] ?? []
            for row in structure where isDueStructureItem(row) {
                guard let due = LMSDates.parse(row.dueAt) else { continue }
                events.append(
                    PlannerCalendarEvent(
                        id: "due:\(course.courseCode):\(row.id)",
                        title: row.title,
                        courseCode: course.courseCode,
                        courseTitle: course.displayTitle,
                        startsAt: due,
                        endsAt: nil,
                        allDay: false,
                        kind: calendarKind(for: row.kind),
                        structureKind: row.kind,
                        structureItemId: row.id,
                        notebookPageId: nil
                    )
                )
            }
        }

        for task in notebookTasks where !task.completed {
            guard let dueRaw = task.dueAt, let due = LMSDates.parse(dueRaw) else { continue }
            events.append(
                PlannerCalendarEvent(
                    id: "notebook:\(task.id)",
                    title: task.taskText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
                        ? "Notebook task"
                        : task.taskText.trimmingCharacters(in: .whitespacesAndNewlines),
                    courseCode: task.courseCode == "__global__" ? nil : task.courseCode,
                    courseTitle: task.courseCode == "__global__"
                        ? nil
                        : courseTitle(for: task.courseCode, titles: courseTitles),
                    startsAt: due,
                    endsAt: nil,
                    allDay: false,
                    kind: .notebookTask,
                    structureKind: nil,
                    structureItemId: nil,
                    notebookPageId: task.notebookPageId
                )
            )
        }

        for event in academicEvents {
            guard let start = LMSDates.parse(event.startDate) else { continue }
            let end = event.endDate.flatMap { LMSDates.parse($0) }
            events.append(
                PlannerCalendarEvent(
                    id: "academic:\(event.id)",
                    title: event.eventName,
                    courseCode: nil,
                    courseTitle: nil,
                    startsAt: start,
                    endsAt: end,
                    allDay: event.allDay,
                    kind: .academic,
                    structureKind: nil,
                    structureItemId: nil,
                    notebookPageId: nil
                )
            )
        }

        events.append(contentsOf: OfficeHoursLogic.collectCalendarEvents(
            studentCourses: studentCourses,
            availabilityByCourseCode: officeHoursByCourseCode
        ))

        return events.sorted { $0.startsAt < $1.startsAt }
    }

    static func monthGridCells(monthAnchor: Date) -> [Date] {
        var calendar = Calendar.current
        calendar.firstWeekday = 2
        let start = calendar.date(from: calendar.dateComponents([.year, .month], from: monthAnchor)) ?? monthAnchor
        let mondayOffset = (calendar.component(.weekday, from: start) + 5) % 7
        let gridStart = calendar.date(byAdding: .day, value: -mondayOffset, to: start) ?? start
        return (0 ..< 42).compactMap { calendar.date(byAdding: .day, value: $0, to: gridStart) }
    }

    static func dateKeyLocal(_ date: Date) -> String {
        let calendar = Calendar.current
        let year = calendar.component(.year, from: date)
        let month = calendar.component(.month, from: date)
        let day = calendar.component(.day, from: date)
        return String(format: "%04d-%02d-%02d", year, month, day)
    }

    static func events(on day: Date, events: [PlannerCalendarEvent]) -> [PlannerCalendarEvent] {
        let key = dateKeyLocal(day)
        return events.filter { dateKeyLocal($0.startsAt) == key }
    }

    static func eventCountsByDay(events: [PlannerCalendarEvent]) -> [String: Int] {
        var counts: [String: Int] = [:]
        for event in events {
            let key = dateKeyLocal(event.startsAt)
            counts[key, default: 0] += 1
        }
        return counts
    }

    static func dueSoonItems(from items: [StudentTodoItem], limit: Int = 5, now: Date = Date()) -> [StudentTodoItem] {
        items
            .filter { !$0.isCompleted }
            .filter { item in
                guard let due = item.dueAt else { return false }
                let buckets = bucketTodos([item], now: now)
                return buckets[.overdue]?.isEmpty == false
                    || buckets[.today]?.isEmpty == false
                    || buckets[.thisWeek]?.isEmpty == false
            }
            .prefix(limit)
            .map { $0 }
    }

    static func encodeSnapshot(todos: [StudentTodoItem], events: [PlannerCalendarEvent], fetchedAt: Date = Date()) -> PlannerSnapshot {
        PlannerSnapshot(
            fetchedAt: fetchedAt,
            todos: todos.map(cachedTodo),
            events: events.map(cachedEvent)
        )
    }

    static func decodeSnapshot(_ snapshot: PlannerSnapshot) -> (todos: [StudentTodoItem], events: [PlannerCalendarEvent]) {
        (
            snapshot.todos.map(decodedTodo),
            snapshot.events.map(decodedEvent)
        )
    }

    // MARK: - Private

    private static func isDueStructureItem(_ item: CourseStructureItem) -> Bool {
        (item.kind == "content_page" || item.kind == "assignment" || item.kind == "quiz")
            && item.dueAt?.isEmpty == false
    }

    private static func calendarKind(for structureKind: String) -> PlannerCalendarEventKind {
        switch structureKind {
        case "assignment": return .assignment
        case "quiz": return .quiz
        default: return .contentPage
        }
    }

    private static func completionStatus(itemId: String, kind: String, grades: MyGradesResponse?) -> StudentTodoCompletion {
        guard let grades else { return .open }
        if grades.grades[itemId] != nil || grades.displayGrades[itemId] != nil {
            return .completed
        }
        let status = grades.gradeStatuses[itemId]?.lowercased() ?? ""
        if status.contains("submit") { return .submitted }
        if status.contains("complete") || status.contains("graded") { return .completed }
        return .open
    }

    private static func courseTitle(for courseCode: String, titles: [String: String]) -> String {
        if courseCode == "__global__" { return "Notebook" }
        return titles[courseCode] ?? courseCode
    }

    private static func cachedTodo(_ item: StudentTodoItem) -> CachedStudentTodoItem {
        CachedStudentTodoItem(
            key: item.key,
            kind: item.kind.rawValue,
            title: item.title,
            courseCode: item.courseCode,
            courseTitle: item.courseTitle,
            dueAt: item.dueAt.map { ISO8601DateFormatter().string(from: $0) },
            structureKind: item.structureKind,
            structureItemId: item.structureItemId,
            notebookPageId: item.notebookPageId,
            notebookTaskId: item.notebookTaskId,
            completion: item.completion.rawValue
        )
    }

    private static func decodedTodo(_ cached: CachedStudentTodoItem) -> StudentTodoItem {
        StudentTodoItem(
            key: cached.key,
            kind: StudentTodoKind(rawValue: cached.kind) ?? .dueItem,
            title: cached.title,
            courseCode: cached.courseCode,
            courseTitle: cached.courseTitle,
            dueAt: LMSDates.parse(cached.dueAt),
            structureKind: cached.structureKind,
            structureItemId: cached.structureItemId,
            notebookPageId: cached.notebookPageId,
            notebookTaskId: cached.notebookTaskId,
            completion: StudentTodoCompletion(rawValue: cached.completion) ?? .open
        )
    }

    private static func cachedEvent(_ event: PlannerCalendarEvent) -> CachedPlannerCalendarEvent {
        CachedPlannerCalendarEvent(
            id: event.id,
            title: event.title,
            courseCode: event.courseCode,
            courseTitle: event.courseTitle,
            startsAt: ISO8601DateFormatter().string(from: event.startsAt),
            endsAt: event.endsAt.map { ISO8601DateFormatter().string(from: $0) },
            allDay: event.allDay,
            kind: event.kind.rawValue,
            structureKind: event.structureKind,
            structureItemId: event.structureItemId,
            notebookPageId: event.notebookPageId,
            officeHoursSlotId: event.officeHoursSlotId,
            meetingId: event.meetingId
        )
    }

    private static func decodedEvent(_ cached: CachedPlannerCalendarEvent) -> PlannerCalendarEvent {
        PlannerCalendarEvent(
            id: cached.id,
            title: cached.title,
            courseCode: cached.courseCode,
            courseTitle: cached.courseTitle,
            startsAt: LMSDates.parse(cached.startsAt) ?? Date(),
            endsAt: cached.endsAt.flatMap { LMSDates.parse($0) },
            allDay: cached.allDay,
            kind: PlannerCalendarEventKind(rawValue: cached.kind) ?? .academic,
            structureKind: cached.structureKind,
            structureItemId: cached.structureItemId,
            notebookPageId: cached.notebookPageId,
            officeHoursSlotId: cached.officeHoursSlotId,
            meetingId: cached.meetingId
        )
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/PlannerLogic.kt
package com.lextures.android.core.lms

import java.time.DayOfWeek
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.temporal.TemporalAdjusters
import kotlinx.serialization.Serializable

enum class StudentTodoKind { DueItem, NotebookTask }

enum class StudentTodoCompletion { Open, Submitted, Completed }

enum class StudentTodoBucket { Overdue, Today, ThisWeek, Later }

enum class PlannerCalendarEventKind {
    Assignment,
    Quiz,
    ContentPage,
    NotebookTask,
    Academic,
}

data class StudentTodoItem(
    val key: String,
    val kind: StudentTodoKind,
    val title: String,
    val courseCode: String,
    val courseTitle: String,
    val dueAt: Instant?,
    val structureKind: String? = null,
    val structureItemId: String? = null,
    val notebookPageId: String? = null,
    val notebookTaskId: String? = null,
    val completion: StudentTodoCompletion = StudentTodoCompletion.Open,
) {
    val isCompleted: Boolean
        get() = completion == StudentTodoCompletion.Completed || completion == StudentTodoCompletion.Submitted
}

data class PlannerCalendarEvent(
    val id: String,
    val title: String,
    val courseCode: String? = null,
    val courseTitle: String? = null,
    val startsAt: Instant,
    val endsAt: Instant? = null,
    val allDay: Boolean = false,
    val kind: PlannerCalendarEventKind,
    val structureKind: String? = null,
    val structureItemId: String? = null,
    val notebookPageId: String? = null,
)

data class PlannerCourseFilter(
    val courseCode: String,
    val title: String,
)

@Serializable
data class PlannerSnapshot(
    val fetchedAtEpochMs: Long,
    val todos: List<CachedStudentTodoItem> = emptyList(),
    val events: List<CachedPlannerCalendarEvent> = emptyList(),
)

@Serializable
data class CachedStudentTodoItem(
    val key: String,
    val kind: String,
    val title: String,
    val courseCode: String,
    val courseTitle: String,
    val dueAt: String? = null,
    val structureKind: String? = null,
    val structureItemId: String? = null,
    val notebookPageId: String? = null,
    val notebookTaskId: String? = null,
    val completion: String = StudentTodoCompletion.Open.name,
)

@Serializable
data class CachedPlannerCalendarEvent(
    val id: String,
    val title: String,
    val courseCode: String? = null,
    val courseTitle: String? = null,
    val startsAt: String,
    val endsAt: String? = null,
    val allDay: Boolean = false,
    val kind: String,
    val structureKind: String? = null,
    val structureItemId: String? = null,
    val notebookPageId: String? = null,
)

enum class DueReminderLeadTime(val minutes: Int) {
    None(0),
    FifteenMinutes(15),
    OneHour(60),
    OneDay(1440),
}

object PlannerLogic {
    private const val GLOBAL_NOTEBOOK_KEY = "__global__"

    fun dueItemKey(courseCode: String, itemId: String): String = "due:$courseCode:$itemId"

    fun notebookTaskKey(taskId: String): String = "notebook:$taskId"

    fun collectTodos(
        studentCourses: List<CourseSummary>,
        structureByCourseCode: Map<String, List<CourseStructureItem>>,
        notebookTasks: List<NotebookTask>,
        gradesByCourseCode: Map<String, MyGradesResponse>,
    ): List<StudentTodoItem> {
        val courseTitles = studentCourses.associate { it.courseCode to it.displayTitle }
        val studentCodes = studentCourses.map { it.courseCode }.toSet()
        val items = mutableListOf<StudentTodoItem>()

        for (task in notebookTasks) {
            if (task.completed) continue
            if (task.courseCode != GLOBAL_NOTEBOOK_KEY && task.courseCode !in studentCodes) continue
            val title = task.taskText.trim().ifEmpty { "Untitled task" }
            items += StudentTodoItem(
                key = notebookTaskKey(task.id),
                kind = StudentTodoKind.NotebookTask,
                title = title,
                courseCode = task.courseCode,
                courseTitle = courseTitle(task.courseCode, courseTitles),
                dueAt = LmsDates.parse(task.dueAt),
                notebookPageId = task.notebookPageId,
                notebookTaskId = task.id,
                completion = if (task.completed) StudentTodoCompletion.Completed else StudentTodoCompletion.Open,
            )
        }

        for (course in studentCourses) {
            if (!course.isCalendarEnabled) continue
            val structure = structureByCourseCode[course.courseCode].orEmpty()
            val grades = gradesByCourseCode[course.courseCode]
            for (row in structure) {
                if (!isDueStructureItem(row)) continue
                items += StudentTodoItem(
                    key = dueItemKey(course.courseCode, row.id),
                    kind = StudentTodoKind.DueItem,
                    title = row.title.trim().ifEmpty { "Untitled" },
                    courseCode = course.courseCode,
                    courseTitle = course.displayTitle,
                    dueAt = LmsDates.parse(row.dueAt),
                    structureKind = row.kind,
                    structureItemId = row.id,
                    completion = completionStatus(row.id, grades),
                )
            }
        }

        return items.sortedWith(compareBy(nullsLast()) { it.dueAt })
    }

    fun bucketTodos(items: List<StudentTodoItem>, zone: ZoneId = ZoneId.systemDefault()): Map<StudentTodoBucket, List<StudentTodoItem>> {
        val now = Instant.now()
        val today = LocalDate.now(zone)
        val startOfToday = today.atStartOfDay(zone).toInstant()
        val endOfToday = today.plusDays(1).atStartOfDay(zone).minusSeconds(1).toInstant()
        val weekStart = today.with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY)).atStartOfDay(zone).toInstant()
        val weekEnd = weekStart.plusSeconds(7 * 86_400 - 1)

        val buckets = StudentTodoBucket.entries.associateWith { mutableListOf<StudentTodoItem>() }
        for (item in items) {
            val due = item.dueAt
            val bucket = when {
                due == null -> StudentTodoBucket.Later
                due < startOfToday -> StudentTodoBucket.Overdue
                due <= endOfToday -> StudentTodoBucket.Today
                due <= weekEnd -> StudentTodoBucket.ThisWeek
                else -> StudentTodoBucket.Later
            }
            buckets.getValue(bucket) += item
        }
        return buckets
    }

    fun collectCalendarEvents(
        studentCourses: List<CourseSummary>,
        structureByCourseCode: Map<String, List<CourseStructureItem>>,
        notebookTasks: List<NotebookTask>,
        academicEvents: List<AcademicCalendarEvent>,
    ): List<PlannerCalendarEvent> {
        val courseTitles = studentCourses.associate { it.courseCode to it.displayTitle }
        val events = mutableListOf<PlannerCalendarEvent>()

        for (course in studentCourses) {
            if (!course.isCalendarEnabled) continue
            for (row in structureByCourseCode[course.courseCode].orEmpty()) {
                if (!isDueStructureItem(row)) continue
                val due = LmsDates.parse(row.dueAt) ?: continue
                events += PlannerCalendarEvent(
                    id = "due:${course.courseCode}:${row.id}",
                    title = row.title,
                    courseCode = course.courseCode,
                    courseTitle = course.displayTitle,
                    startsAt = due,
                    kind = calendarKind(row.kind),
                    structureKind = row.kind,
                    structureItemId = row.id,
                )
            }
        }

        for (task in notebookTasks) {
            if (task.completed) continue
            val due = LmsDates.parse(task.dueAt) ?: continue
            events += PlannerCalendarEvent(
                id = "notebook:${task.id}",
                title = task.taskText.trim().ifEmpty { "Notebook task" },
                courseCode = task.courseCode.takeIf { it != GLOBAL_NOTEBOOK_KEY },
                courseTitle = if (task.courseCode == GLOBAL_NOTEBOOK_KEY) null
                else courseTitle(task.courseCode, courseTitles),
                startsAt = due,
                kind = PlannerCalendarEventKind.NotebookTask,
                notebookPageId = task.notebookPageId,
            )
        }

        for (event in academicEvents) {
            val start = LmsDates.parse(event.startDate) ?: continue
            events += PlannerCalendarEvent(
                id = "academic:${event.id}",
                title = event.eventName,
                startsAt = start,
                endsAt = event.endDate?.let { LmsDates.parse(it) },
                allDay = event.allDay,
                kind = PlannerCalendarEventKind.Academic,
            )
        }

        return events.sortedBy { it.startsAt }
    }

    fun monthGridCells(monthAnchor: LocalDate, zone: ZoneId = ZoneId.systemDefault()): List<LocalDate> {
        val start = monthAnchor.withDayOfMonth(1)
        val mondayOffset = (start.dayOfWeek.value + 6) % 7
        val gridStart = start.minusDays(mondayOffset.toLong())
        return (0 until 42).map { gridStart.plusDays(it.toLong()) }
    }

    fun dateKeyLocal(instant: Instant, zone: ZoneId = ZoneId.systemDefault()): String =
        instant.atZone(zone).toLocalDate().toString()

    fun dateKeyLocal(date: LocalDate): String = date.toString()

    fun eventsOnDay(day: LocalDate, events: List<PlannerCalendarEvent>, zone: ZoneId = ZoneId.systemDefault()): List<PlannerCalendarEvent> {
        val key = day.toString()
        return events.filter { dateKeyLocal(it.startsAt, zone) == key }
    }

    fun eventCountsByDay(events: List<PlannerCalendarEvent>, zone: ZoneId = ZoneId.systemDefault()): Map<String, Int> =
        events.groupingBy { dateKeyLocal(it.startsAt, zone) }.eachCount()

    fun dueSoonItems(items: List<StudentTodoItem>, limit: Int = 5): List<StudentTodoItem> {
        val buckets = bucketTodos(items.filter { !it.isCompleted })
        return (buckets[StudentTodoBucket.Overdue].orEmpty() +
            buckets[StudentTodoBucket.Today].orEmpty() +
            buckets[StudentTodoBucket.ThisWeek].orEmpty())
            .take(limit)
    }

    fun encodeSnapshot(todos: List<StudentTodoItem>, events: List<PlannerCalendarEvent>): PlannerSnapshot =
        PlannerSnapshot(
            fetchedAtEpochMs = System.currentTimeMillis(),
            todos = todos.map(::cachedTodo),
            events = events.map(::cachedEvent),
        )

    fun decodeSnapshot(snapshot: PlannerSnapshot): Pair<List<StudentTodoItem>, List<PlannerCalendarEvent>> =
        snapshot.todos.map(::decodedTodo) to snapshot.events.map(::decodedEvent)

    private fun isDueStructureItem(item: CourseStructureItem): Boolean =
        item.kind in setOf("content_page", "assignment", "quiz") && !item.dueAt.isNullOrBlank()

    private fun calendarKind(structureKind: String): PlannerCalendarEventKind = when (structureKind) {
        "assignment" -> PlannerCalendarEventKind.Assignment
        "quiz" -> PlannerCalendarEventKind.Quiz
        else -> PlannerCalendarEventKind.ContentPage
    }

    private fun completionStatus(itemId: String, grades: MyGradesResponse?): StudentTodoCompletion {
        if (grades == null) return StudentTodoCompletion.Open
        if (grades.grades[itemId] != null || grades.displayGrades[itemId] != null) {
            return StudentTodoCompletion.Completed
        }
        val status = grades.gradeStatuses[itemId]?.lowercase().orEmpty()
        if (status.contains("submit")) return StudentTodoCompletion.Submitted
        if (status.contains("complete") || status.contains("graded")) return StudentTodoCompletion.Completed
        return StudentTodoCompletion.Open
    }

    private fun courseTitle(courseCode: String, titles: Map<String, String>): String =
        if (courseCode == GLOBAL_NOTEBOOK_KEY) "Notebook" else titles[courseCode] ?: courseCode

    private fun cachedTodo(item: StudentTodoItem) = CachedStudentTodoItem(
        key = item.key,
        kind = item.kind.name,
        title = item.title,
        courseCode = item.courseCode,
        courseTitle = item.courseTitle,
        dueAt = item.dueAt?.toString(),
        structureKind = item.structureKind,
        structureItemId = item.structureItemId,
        notebookPageId = item.notebookPageId,
        notebookTaskId = item.notebookTaskId,
        completion = item.completion.name,
    )

    private fun decodedTodo(cached: CachedStudentTodoItem) = StudentTodoItem(
        key = cached.key,
        kind = runCatching { StudentTodoKind.valueOf(cached.kind) }.getOrDefault(StudentTodoKind.DueItem),
        title = cached.title,
        courseCode = cached.courseCode,
        courseTitle = cached.courseTitle,
        dueAt = LmsDates.parse(cached.dueAt),
        structureKind = cached.structureKind,
        structureItemId = cached.structureItemId,
        notebookPageId = cached.notebookPageId,
        notebookTaskId = cached.notebookTaskId,
        completion = runCatching { StudentTodoCompletion.valueOf(cached.completion) }
            .getOrDefault(StudentTodoCompletion.Open),
    )

    private fun cachedEvent(event: PlannerCalendarEvent) = CachedPlannerCalendarEvent(
        id = event.id,
        title = event.title,
        courseCode = event.courseCode,
        courseTitle = event.courseTitle,
        startsAt = event.startsAt.toString(),
        endsAt = event.endsAt?.toString(),
        allDay = event.allDay,
        kind = event.kind.name,
        structureKind = event.structureKind,
        structureItemId = event.structureItemId,
        notebookPageId = event.notebookPageId,
    )

    private fun decodedEvent(cached: CachedPlannerCalendarEvent) = PlannerCalendarEvent(
        id = cached.id,
        title = cached.title,
        courseCode = cached.courseCode,
        courseTitle = cached.courseTitle,
        startsAt = LmsDates.parse(cached.startsAt) ?: Instant.now(),
        endsAt = cached.endsAt?.let { LmsDates.parse(it) },
        allDay = cached.allDay,
        kind = runCatching { PlannerCalendarEventKind.valueOf(cached.kind) }
            .getOrDefault(PlannerCalendarEventKind.Academic),
        structureKind = cached.structureKind,
        structureItemId = cached.structureItemId,
        notebookPageId = cached.notebookPageId,
    )
}

val CourseSummary.isCalendarEnabled: Boolean
    get() = calendarEnabled != false
